### ISSUE TITLE
target kernel code generation

### ISSUE THREAD

**Issue by yanyh15** [passlab/rexompiler#79]
OpenMP `target` with nested parallel/for could be very complicated: 

```
#pragma omp target // (1 team and 1 thread per team)
{
printf("position 1: master team, and master thread %d/%d, %d/%d\n", omp_get_team_num(), omp_get_num_teams(), omp_get_thread_num(), omp_get_num_threads());

#pragma omp parallel num_threads(16)// (1 team and 16 threads) 

{

printf("position 2: team, and thread %d/%d, %d/%d\n", omp_get_team_num(), omp_get_num_teams(), omp_get_thread_num(), omp_get_num_threads());

}

...

#pragma omp teams parallel num_teams (4) num_threads(8) //(4 teams and each has 8 threads)
{

printf("position 3: team, and thread %d/%d, %d/%d\n", omp_get_team_num(), omp_get_num_teams(), omp_get_thread_num(), omp_get_num_threads());

}

....

}
```

**Comment by XinyaoYI** [comment]
CUDA code:
```
#include <cuda_runtime.h>
#include <stdio.h>
__global__ void kernel3()
{
    //unsigned int tid=blockIdx.x*blockDim.x+threadIdx.x;
    unsigned int tid=threadIdx.x;//get the index in the block

    printf("position3:team, and thread. block_id: %d, thread_id:%d\n",blockIdx.x, tid);
    // #pragma omp for schedule(dynamic:4)
}


__global__ void kernel2()
{
    //unsigned int tid=blockIdx.x*blockDim.x+threadIdx.x;
    unsigned int tid=threadIdx.x;
    printf("position2:team, and thread. block_id: %d, thread_id:%d\n",blockIdx.x, tid);

}


__global__ void kernel1()
{
    unsigned int tid=blockIdx.x*blockDim.x+threadIdx.x;
    printf("position1:master team(block), and master thread(thread). block_id: %d, thread_id:%d\n",blockIdx.x, tid);
    kernel2<<<1,16>>>();
    kernel3<<<4,8>>>();


}

int main(int argc,char* argv[])
{
    kernel1<<<1,1>>>();
    cudaGetLastError();
    cudaDeviceReset();
    return 0;
}
```

**Comment by yanyh15** [comment]
LLVM uses state machine, i.e. the root kernel starts with max-reasonable number of blocks/threads, then internall use state machine to put the threads that are not needed in waiting unti later.

**Comment by yanyh15** [comment]
1. Overhead evaluation of using dynamic parallelism using AXPY, MV, MM. measurement of performance of using and not using dynamic parallelism and plot the results in bar chart: X is problem size, Y is for the performance. For each problem size, there are two (or more if needed) bars to show the performance of using and not using dynamic parallelism
2. Evaluation of dynamic parallelism using image processing pipeline (https://github.com/passlab/ImagePipeline)
3. Implement a CNN network using OpenMP target, the skeleton code would be like this: 

       #pragma omp target data map (...)    //map data
       while ( ... ) //gradient descent {
                 #pragma omp target 
                {
                      //convolution 1
                      //pooling 1

                      //convolution 2
                     //pooling 2

                     ....

                    //softmax
             }
     
      }
